phabricatoremails/template.py

```diff
@@ -0,0 +1,237 @@
+import textwrap
+from datetime import datetime, timezone
+
+import jinja2
+from phabricatoremails.model.common import Reviewer, ReviewerStatus
+from phabricatoremails.model.phabricator import (
+    ReplyContext,
+    CodeContext,
+    DiffLineType,
+    AffectedFileChange,
+)
+from phabricatoremails.model.phabricator_secure import ExistenceChange
+from premailer import Premailer
+
+
+def is_reply(context):
+    return isinstance(context, ReplyContext)
+
+
+def is_code(context):
+    return isinstance(context, CodeContext)
+
+
+def is_has_comments(event):
+    return event.main_comment or event.inline_comments
+
+
+def is_has_comments_secure(event):
+    return event.comment_count > 0
+
+
+def is_reviewer_accepted(reviewer: Reviewer):
+    return not reviewer.is_actionable and reviewer.status == ReviewerStatus.ACCEPTED
+
+
+def is_reviewer_blocking(reviewer: Reviewer):
+    return reviewer.status == ReviewerStatus.BLOCKING
+
+
+def is_reviewer_soft_needed(reviewer: Reviewer):
+    # The revision needs a review from _someone_, and this person _is_ a reviewer, but
+    # it's not them specifically who needs to perform the review
+    return reviewer.is_actionable and reviewer.status == ReviewerStatus.UNREVIEWED
+
+
+def is_reviewer_needed(reviewer: Reviewer):
+    # This specific reviewer needs to accept this revision before it can be landed
+    return reviewer.is_actionable and reviewer.status in [
+        ReviewerStatus.REQUESTED_CHANGES,
+        ReviewerStatus.BLOCKING,
+    ]
+
+
+def reviewer_actionable_class(reviewer: Reviewer):
+    return "actionable" if reviewer.is_actionable else "non-actionable"
+
+
+def date(utc_value: datetime, receiver_timezone: timezone):
+    return utc_value.astimezone(receiver_timezone).strftime("%b %-d %-I:%M%p:")
+
+
+def comment_summary(body):
+    if not body.main_comment and not body.inline_comments:
+        return ""
+    elif (body.main_comment and not body.inline_comments) or (
+        not body.main_comment and len(body.inline_comments) == 1
+    ):
+        return " and submitted a comment"
+    else:
+        return " and submitted comments"
+
+
+def secure_comment_summary(body):
+    if body.comment_count == 0:
+        return ""
+    elif body.comment_count == 1:
+        return " and submitted a comment"
+    else:
+        return " and submitted comments"
+
+
+def diff_class(diff_type: DiffLineType):
+    if diff_type == DiffLineType.ADDED:
+        return "added-line"
+    elif diff_type == DiffLineType.REMOVED:
+        return "removed-line"
+    elif diff_type == DiffLineType.NO_CHANGE:
+        return "no-change-line"
+
+
+def reviewer_status_icon(status: ReviewerStatus):
+    if status == ReviewerStatus.ACCEPTED:
+        return "&#x2714;"
```

Comment:
This as well as all other instances of similar unicode characters being used should be put in an enum or dictionary and given human readable names, e.g. `UNICODE_CHECKMARK`.